### PR TITLE
Add AVFoundation support for OS X, including support for CAP_PROP_MODE

### LINKED ISSUE #4930: Add avfoundation support for normal OSX build, not just iOS
Transferred from http://code.opencv.org/issues/4187
```
|| Dustin Spicuzza on 2015-02-10 06:01
|| Priority: Normal
|| Affected: branch 'master' (3.0-dev)
|| Category: highgui-video
|| Tracker: Bug
|| Difficulty: Easy
|| PR: 
|| Platform: x64 / Mac OSX
```

Add avfoundation support for normal OSX build, not just iOS
-----------
```
When using QTKit, things like exposure settings and such aren't supported. AVFoundation supports doing this, however.

I hacked with the makefiles in a very minimal way, and I found the following:

* Can't enable QTKit and AVFoundation at the same time, otherwise link errors occur
* Have to add some frameworks in CMakeLists.txt in videoio -- here's what my line that worked was (not sure if there's too many frameworks there, I just copy/pasted settings from IOS):

<pre>
list(APPEND VIDEOIO_LIBRARIES "-framework Accelerate" "-framework AVFoundation" "-framework CoreGraphics" "-framework CoreMedia" "-framework CoreVideo" )
</pre>

Once I did this very minimal change, I compiled and was able to set the exposure on the webcam using PROP_IOS_DEVICE_EXPOSURE on my Macbook Pro. I haven't done extensive testing yet, but my understanding is that AVFoundation is mostly identical on iOS and OSX.
```

History
-------
##### Vadim Pisarevsky on 2015-05-13 13:28
```
-   Category changed from imgproc, video to highgui-video
```

### PR DESCRIPTION
**Pull request by mself**
resolves #4930 

Extends the support for AVFoundation on Mac that was started in PR #6514.  This PR also includes support for CAP_PROP_MODE, which lets you get grayscale frames 2x faster than before.

To use:
        capture.open(filename, CV_CAP_AVFOUNDATION_MAC);
        capture.set(CAP_PROP_MODE, CV_CAP_MODE_GRAY);

### CODE REVIEW

**Comment by yoffy** [comment]
It sounds and change set seems very cool! :+1:

**Comment by paroj** [comment]
I would prefer the `CAP_PROP_CONVERT_RGB` semantics over `CAP_PROP_MODE` as it moves the conversion logic out of the individual VideoCapture implementations. See #5485.

Implementation of V4L: https://github.com/Itseez/opencv/blob/master/modules/videoio/src/cap_v4l.cpp#L1761

**Comment by mself** [comment]
Hi @paroj, I looked at that code, but I don't see how that would let me return a grayscale image without copying the color information, too.  How would the VideoCapture know that I only want grayscale?  I also read through #5485 and I don't think CAP_PROP_CONVERT_RGB is a sufficient API.

Let me start by being clear about my goals/requirements.  I'm working on a real-time video processing that only needs grayscale data.  I'm reading from video files (not cameras), so the incoming data is typically in 420 format (H.264) or possibly some other YUV format.  The source doesn't have a "native" format, since it all depends on what video file was opened.

In order to achieve real-time processing speeds, it's important to avoid copying the chroma data at all.  The performance drop is very significant.  So, there needs to be some way for the caller to indicate that they only want grayscale data to be returned.

capture.set(CAP_PROP_MODE, CV_CAP_MODE_GRAY) seems to do this pretty clearly (if we just change the docs to be clearer about what it does by deleting the phrase 'Backend-specific value").  The docs should say that this is the color format that the caller wants.  The API can return false if it isn't able to do it (or it would be just as efficient for the caller to the conversion themselves).

I am OK dropping CV_CAP_MODE_RGB -- that seems like a distraction.  If you want RGB data then you should probably use OpenCV's BGR standard format.  And I don't have any opinion on whether CV_CAP_MODE_YUYV should be supported.  Does anyone use it?

One of the issues raised in #5485 was pushing complexity into the VideoCapture implementations.  I did not find this to be a problem.  The underlying AVFoundation code can decode video into one of several formats, so it was just a matter of creating a mapping from the desired OCV format to the fastest AVF option plus an appropriate call to cvtColor() to move the data.  I didn't do anything weird like memcpy() or return references to any internal buffers (the data is always copied once).

After testing all of the options, I found basically the same result that cap_v4l did:
- For BGR data, let AVF do the conversion to ABGR and copy the BGR data out
- For grayscale, let AVF decode to 420 and copy the Y data only (much faster than decoding to 422)
- For YUYU, let AVF decode to 422 and copy al of the data out

It's not that much code.  And some of it could probably be pushed up into the base class, if it gets repeated too much.

I'd be interested in comments on these possible changes to improve the commit:
1) Remove support for CV_CAP_MODE_RGB (in favor of CV_CAP_MODE_BGR, the default)
2) Remove support for CV_CAP_MODE_YUYV
3) Update docs to clarify semantics of CAP_PROP_MODE

**Comment by paroj** [comment]
> 1) Remove support for CV_CAP_MODE_RGB (in favor of CV_CAP_MODE_BGR, the default)
2) Remove support for CV_CAP_MODE_YUYV

I agree with these two. However this makes `CV_CAP_MODE_GRAY` the only sensible option for `CAP_PROP_MODE`.
Therefore I would rather introduce `CAP_PROP_MODE_GRAYSCALE` boolean property similar to `CAP_PROP_CONVERT_RGB`.

This would keep the current `CAP_PROP_MODE` semantics which are implemented by most capture backends:
- the `cap_libv4l` implementation interprets it as "output mode" and  returns YUYV images.
- the `cap_*1394*` implementations interpret it as "transmission mode" and always return BGR images. (converting gray to BGR)
- the `cap_v4l` implementation also interprets it as "transmission mode", but treats it read only
- other `cap_*` implementations do not support `CAP_PROP_MODE`

(updated #5485 accordingly)

File: modules/videoio/src/cap_avfoundation_mac.mm [hunk 1]
```diff
@@ -0,0 +1,1314 @@
...
+    mDeviceImage->width = int(width);
+    mDeviceImage->height = int(height);
+    mDeviceImage->nChannels = deviceChannels;
+    mDeviceImage->depth = IPL_DEPTH_8U;
+    mDeviceImage->widthStep = int(rowBytes);
+    mDeviceImage->imageData = reinterpret_cast<char *>(baseaddress);
+    mDeviceImage->imageSize = int(rowBytes*height);
+
+    // Convert the device image into the output image.
+    if (cvtCode == 0) {
```
Comment (yoffy) [inline modules/videoio/src/cap_avfoundation_mac.mm]: `cvtCode == 0` means `CV_BGR2BGRA`.
In the future, I think there is a possibility that `pixelFormat == kCVPixelFormatType_24BGR && mMode == CV_CAP_MODE_BGRA`.

CV_BGR2BGRA:
http://docs.opencv.org/3.1.0/df/d4e/group__imgproc__c.html#enum-members

kCVPixelFormatType_24BGR (in Pixel Format Types):
https://developer.apple.com/library/mac/documentation/QuartzCore/Reference/CVPixelFormatDescriptionRef/

File: modules/videoio/src/cap_avfoundation_mac.mm [hunk 2]
```diff
@@ -0,0 +1,1314 @@
...
+
+    // create output
+    mCapture = [[CaptureDelegate alloc] init];
+    mCaptureVideoDataOutput = [[AVCaptureVideoDataOutput alloc] init];
+    dispatch_queue_t queue = dispatch_queue_create("cameraQueue", DISPATCH_QUEUE_SERIAL);
+    [mCaptureVideoDataOutput setSampleBufferDelegate: mCapture queue: queue];
+    dispatch_release(queue);
+
+    OSType pixelFormat = kCVPixelFormatType_32BGRA;
+    //OSType pixelFormat = kCVPixelFormatType_422YpCbCr8;
```
Comment (yoffy) [inline modules/videoio/src/cap_avfoundation_mac.mm]: I think you can make to same behaviour as `CvCaptureFile::setupReadingAt`.

**Comment by vpisarev** [comment]
the PR will be reopened because of CI issues

**Pull request closed without merging**